File: src/api/sources/mangadex.ts
// MangaDex source (api.mangadex.org) — an open, documented API. Used as a backup
// when atsu.moe is unreachable, or as the primary via `source: "mangadex"`.

import { httpJson } from "../client.ts";
import { readCache, writeCache } from "../../utils/cache.ts";
import { MANGA_CACHE, SEARCH_CACHE, CHAPTERS_CACHE } from "../../utils/paths.ts";
import type {
  BrowseSort,
  DiscoveryItem,
  DiscoveryKind,
  Filters,
  MangaInfo,
  ReadChapter,
  SearchResult,
  Source,
} from "../types.ts";

const API = "https://api.mangadex.org";
const UPLOADS = "https://uploads.mangadex.org";
const UA = "manga-cli/0.3 (+https://github.com/devcheckra1n/manga-cli)";
const HEADERS = { "User-Agent": UA };
const TTL = 60 * 60 * 1000;

type Params = Record<string, string | number | string[] | undefined>;

function url(path: string, params: Params = {}): string {
  const u = new URL(API + path);
  for (const [k, v] of Object.entries(params)) {
    if (v === undefined) continue;
    if (Array.isArray(v)) for (const x of v) u.searchParams.append(k, x);
    else u.searchParams.append(k, String(v));
  }
  return u.toString();
}

function ratings(adult: boolean): string[] {
  return adult ? ["safe", "suggestive", "erotica", "pornographic"] : ["safe", "suggestive"];
}

interface MdManga {
  id: string;
  attributes: {
    title: Record<string, string>;
    altTitles?: Array<Record<string, string>>;
    year?: number | null;
    status?: string;
    contentRating?: string;
    availableTranslatedLanguages?: string[];
    tags?: Array<{ id: string; attributes: { name: Record<string, string>; group: string } }>;
  };
  relationships?: Array<{ type: string; id: string; attributes?: { fileName?: string } }>;
}

function pickTitle(a: MdManga["attributes"]): string {
  return (
    a.title.en ??
    Object.values(a.title)[0] ??
    a.altTitles?.find((t) => t.en)?.en ??
    "Untitled"
  );
}

function coverUrl(m: MdManga): string | undefined {
  const file = m.relationships?.find((r) => r.type === "cover_art")?.attributes?.fileName;
  return file ? `${UPLOADS}/covers/${m.id}/${file}.512.jpg` : undefined;
}

function mapManga(m: MdManga): SearchResult {
  const cr = m.attributes.contentRating ?? "safe";
  const status = m.attributes.status;
  return {
    id: m.id,
    title: pickTitle(m.attributes),
    poster: coverUrl(m),
    type: "Manga",
    status: status ? status[0].toUpperCase() + status.slice(1) : undefined,
    year: m.attributes.year ?? undefined,
    isAdult: cr === "pornographic" || cr === "erotica",
    source: "mangadex",
  };
}

const ORDER: Record<DiscoveryKind, [string, string]> = {
  trending: ["order[followedCount]", "desc"],
  popular: ["order[followedCount]", "desc"],
  mostBookmarked: ["order[followedCount]", "desc"],
  topRated: ["order[rating]", "desc"],
  recentlyAdded: ["order[createdAt]", "desc"],
  recentlyUpdated: ["order[latestUploadedChapter]", "desc"],
};

async function listManga(params: Params): Promise<SearchResult[]> {
  const res = await httpJson<{ data?: MdManga[] }>(
    url("/manga", {
      limit: 24,
      "includes[]": ["cover_art"],
      "availableTranslatedLanguage[]": ["en"], // only titles with English chapters
      ...params,
    }),
    { headers: HEADERS },
  );
  return (res.data ?? []).map(mapManga);
}

interface MdChapter {
  id: string;
  attributes: {
    chapter: string | null;
    title: string | null;
    pages: number;
    translatedLanguage?: string;
    publishAt?: string;
    createdAt?: string;
  };
  relationships?: Array<{ type: string; id: string }>;
}

/** Bubble exact / prefix title matches to the top (MangaDex relevance is weak). */
function rerank(results: SearchResult[], query: string): SearchResult[] {
  const q = query.toLowerCase().trim();
  return results
    .map((r, i) => {
      const t = r.title.toLowerCase();
      let score = i;
      if (t === q) score -= 1000;
      else if (t.startsWith(q)) score -= 500;
      else if (t.includes(q)) score -= 100;
      return { r, score };
    })
    .sort((a, b) => a.score - b.score)
    .map((x) => x.r);
}

// Paginate a manga's feed for the given languages and keep only readable
// (non-external, pages > 0) chapters.
async function fetchReadable(mangaId: string, langs: string[]): Promise<MdChapter[]> {
  const raw: MdChapter[] = [];
  let offset = 0;
  let total = Infinity;
  while (offset < total && offset < 2000) {
    const feed = await httpJson<{ data?: MdChapter[]; total?: number }>(
      url(`/manga/${mangaId}/feed`, {
        "translatedLanguage[]": langs,
        "order[chapter]": "asc",
        "includes[]": ["scanlation_group"],
        "contentRating[]": ["safe", "suggestive", "erotica", "pornographic"],
        limit: 100,
        offset,
      }),
      { headers: HEADERS },
    );
    raw.push(...(feed.data ?? []));
    total = feed.total ?? raw.length;
    offset += 100;
    if (!feed.data || feed.data.length === 0) break;
  }
  return raw.filter((ch) => (ch.attributes.pages ?? 0) > 0);
}

export const mangadex: Source = {
  id: "mangadex",
  label: "MangaDex (mangadex.org)",
  available: true,

  async search(query, opts) {
    const key = `md:search:${query}:${opts.adult ? 1 : 0}:${opts.page ?? 0}`;
    const hit = await readCache<SearchResult[]>(SEARCH_CACHE, key, 10 * 60 * 1000);
    if (hit) return hit;
    const results = rerank(
      await listManga({
        title: query,
        offset: (opts.page ?? 0) * 24,
        "contentRating[]": ratings(opts.adult ?? false),
        "hasAvailableChapters": "true", // skip delicensed titles with no readable chapters
        "order[relevance]": "desc",
      }),
      query,
    );
    await writeCache(SEARCH_CACHE, key, results);
    return results;
  },

  async discovery(kind, page, adult): Promise<DiscoveryItem[]> {
    const [orderKey, dir] = ORDER[kind];
    return listManga({
      offset: page * 24,
      "contentRating[]": ratings(adult),
      [orderKey]: dir,
      "hasAvailableChapters": "true",
    });
  },

  async filters(): Promise<Filters> {
    const key = "md:tags";
    const hit = await readCache<Filters>(MANGA_CACHE, key, 24 * 60 * 60 * 1000);
    if (hit) return hit;
    const res = await httpJson<{ data?: Array<{ id: string; attributes: { name: Record<string, string>; group: string } }> }>(
      url("/manga/tag"),
      { headers: HEADERS },
    );
    const genres = (res.data ?? [])
      .filter((t) => t.attributes.group === "genre" || t.attributes.group === "theme")
      .map((t) => ({ id: t.id, name: t.attributes.name.en ?? Object.values(t.attributes.name)[0] }))
      .sort((a, b) => a.name.localeCompare(b.name));
    const filters: Filters = {
      genres,
      tags: [],
      types: [],
      statuses: [
        { id: "ongoing", name: "Ongoing" },
        { id: "completed", name: "Completed" },
        { id: "hiatus", name: "Hiatus" },
        { id: "cancelled", name: "Cancelled" },
      ],
    };
    await writeCache(MANGA_CACHE, key, filters);
    return filters;
  },

  async browse(f): Promise<SearchResult[]> {
    const order: Record<BrowseSort, [string, string]> = {
      popular: ["order[followedCount]", "desc"],
      latest: ["order[latestUploadedChapter]", "desc"],
      rating: ["order[rating]", "desc"],
      alphabetical: ["order[title]", "asc"],
    };
    const [orderKey, dir] = order[f.sort ?? "popular"];
    return listManga({
      "includedTags[]": f.genreId ? [f.genreId] : undefined,
      "status[]": f.status ? [f.status] : undefined,
      "contentRating[]": ratings(f.adult ?? false),
      hasAvailableChapters: "true",
      [orderKey]: dir,
    });
  },

  async info(mangaId): Promise<MangaInfo> {
    const key = `md:info:${mangaId}`;
    const hit = await readCache<MangaInfo>(MANGA_CACHE, key, TTL);
    if (hit) return hit;

    const detail = await httpJson<{ data: MdManga }>(
      url(`/manga/${mangaId}`, { "includes[]": ["cover_art"] }),
      { headers: HEADERS },
    );
    const longStrip = (detail.data.attributes.tags ?? []).some(
      (t) => (t.attributes.name.en ?? "").toLowerCase() === "long strip",
    );

    // Prefer English; if it has no *readable* chapters, fall back to whichever
    // language has the most (so big multi-language titles still open).
    let raw = await fetchReadable(mangaId, ["en"]);
    if (raw.length === 0) {
      const others = (detail.data.attributes.availableTranslatedLanguages ?? []).filter((l) => l && l !== "en");
      if (others.length > 0) {
        const all = await fetchReadable(mangaId, others);
        const byLang = new Map<string, MdChapter[]>();
        for (const ch of all) {
          const l = ch.attributes.translatedLanguage ?? "??";
          const arr = byLang.get(l) ?? [];
          arr.push(ch);
          byLang.set(l, arr);
        }
        for (const arr of byLang.values()) if (arr.length > raw.length) raw = arr;
      }
    }

    // One release per chapter number (first scanlation wins).
    const seen = new Set<string>();
    const picked: MdChapter[] = [];
    for (const ch of raw) {
      const k = ch.attributes.chapter ?? `oneshot:${ch.id}`;
      if (seen.has(k)) continue;
      seen.add(k);
      picked.push(ch);
    }
    picked.sort((a, b) => (parseFloat(a.attributes.chapter ?? "1e9") || 1e9) - (parseFloat(b.attributes.chapter ?? "1e9") || 1e9));

    const info: MangaInfo = {
      id: mangaId,
      title: pickTitle(detail.data.attributes),
      type: "Manga",
      forceStrip: longStrip,
      chapters: picked.map((ch, i) => {
        const num = parseFloat(ch.attributes.chapter ?? "");
        return {
          id: ch.id,
          title: ch.attributes.title || (ch.attributes.chapter ? `Chapter ${ch.attributes.chapter}` : "Oneshot"),
          number: Number.isFinite(num) ? num : i + 1,
          index: i,
          pageCount: ch.attributes.pages ?? 0,
          scanId: ch.relationships?.find((r) => r.type === "scanlation_group")?.id ?? "",
          createdAt: Date.parse(ch.attributes.publishAt ?? ch.attributes.createdAt ?? "") || undefined,
        };
      }),
    };
    await writeCache(MANGA_CACHE, key, info);
    return info;
  },

  async pages(_mangaId, chapterId): Promise<ReadChapter> {
    const key = `md:pages:${chapterId}`;
    const hit = await readCache<ReadChapter>(CHAPTERS_CACHE, key, TTL);
    if (hit) return hit;
    const res = await httpJson<{ baseUrl: string; chapter: { hash: string; data: string[] } }>(
      url(`/at-home/server/${chapterId}`),
      { headers: HEADERS },
    );
    const { baseUrl, chapter } = res;
    const out: ReadChapter = {
      id: chapterId,
      title: "",
      scanId: "",
      pages: chapter.data.map((file, i) => ({
        id: `${chapterId}-${i}`,
        url: `${baseUrl}/data/${chapter.hash}/${file}`,
        number: i,
        width: 0,
        height: 0,
        aspectRatio: 1,
      })),
    };
    await writeCache(CHAPTERS_CACHE, key, out);
    return out;
  },

  // MangaDex has no public "similar" endpoint; recommendations fall back elsewhere.
  async related(): Promise<DiscoveryItem[]> {
    return [];
  },
};

File: go-rewrite/internal/api/mangadex.go
// MangaDex source (api.mangadex.org) — an open, documented JSON API.

package api

import (
	"fmt"
	"net/url"
	"sort"
	"strconv"
	"strings"
	"time"

	"github.com/devcheckra1n/manga-cli/go-rewrite/internal/util"
)

const (
	mdAPI     = "https://api.mangadex.org"
	mdUploads = "https://uploads.mangadex.org"
	mdTTL     = time.Hour
)

var mdHeaders = map[string]string{
	"User-Agent": "manga-cli/2.0 (+https://github.com/devcheckra1n/manga-cli)",
}

type mangadexSource struct{}

func (mangadexSource) ID() SourceID    { return SrcMangaDex }
func (mangadexSource) Label() string   { return "MangaDex (mangadex.org)" }
func (mangadexSource) Available() bool { return true }

func mdURL(path string, params url.Values) string {
	if len(params) == 0 {
		return mdAPI + path
	}
	return mdAPI + path + "?" + params.Encode()
}

func mdRatings(adult bool) []string {
	if adult {
		return []string{"safe", "suggestive", "erotica", "pornographic"}
	}
	return []string{"safe", "suggestive"}
}

// ── API shapes ─────────────────────────────────────────────────────────────────

type mdTag struct {
	ID         string `json:"id"`
	Attributes struct {
		Name  map[string]string `json:"name"`
		Group string            `json:"group"`
	} `json:"attributes"`
}

type mdManga struct {
	ID         string `json:"id"`
	Attributes struct {
		Title                          map[string]string   `json:"title"`
		AltTitles                      []map[string]string `json:"altTitles"`
		Year                           *int                `json:"year"`
		Status                         string              `json:"status"`
		ContentRating                  string              `json:"contentRating"`
		AvailableTranslatedLanguages   []string            `json:"availableTranslatedLanguages"`
		Tags                           []mdTag             `json:"tags"`
	} `json:"attributes"`
	Relationships []struct {
		Type       string `json:"type"`
		ID         string `json:"id"`
		Attributes *struct {
			FileName string `json:"fileName"`
		} `json:"attributes"`
	} `json:"relationships"`
}

type mdChapter struct {
	ID         string `json:"id"`
	Attributes struct {
		Chapter            *string `json:"chapter"`
		Title              *string `json:"title"`
		Pages              int     `json:"pages"`
		TranslatedLanguage string  `json:"translatedLanguage"`
		PublishAt          string  `json:"publishAt"`
		CreatedAt          string  `json:"createdAt"`
	} `json:"attributes"`
	Relationships []struct {
		Type string `json:"type"`
		ID   string `json:"id"`
	} `json:"relationships"`
}

func mdPickTitle(m *mdManga) string {
	if t := m.Attributes.Title["en"]; t != "" {
		return t
	}
	for _, t := range m.Attributes.Title {
		if t != "" {
			return t
		}
	}
	for _, alt := range m.Attributes.AltTitles {
		if t := alt["en"]; t != "" {
			return t
		}
	}
	return "Untitled"
}

func mdCoverURL(m *mdManga) string {
	for _, r := range m.Relationships {
		if r.Type == "cover_art" && r.Attributes != nil && r.Attributes.FileName != "" {
			return fmt.Sprintf("%s/covers/%s/%s.512.jpg", mdUploads, m.ID, r.Attributes.FileName)
		}
	}
	return ""
}

func mdMapManga(m *mdManga) SearchResult {
	cr := m.Attributes.ContentRating
	status := m.Attributes.Status
	if status != "" {
		status = strings.ToUpper(status[:1]) + status[1:]
	}
	year := 0
	if m.Attributes.Year != nil {
		year = *m.Attributes.Year
	}
	return SearchResult{
		ID:      m.ID,
		Title:   mdPickTitle(m),
		Poster:  mdCoverURL(m),
		Type:    "Manga",
		Status:  status,
		Year:    year,
		IsAdult: cr == "pornographic" || cr == "erotica",
		Source:  SrcMangaDex,
	}
}

var mdOrder = map[DiscoveryKind][2]string{
	Trending:        {"order[followedCount]", "desc"},
	Popular:         {"order[followedCount]", "desc"},
	MostBookmarked:  {"order[followedCount]", "desc"},
	TopRated:        {"order[rating]", "desc"},
	RecentlyAdded:   {"order[createdAt]", "desc"},
	RecentlyUpdated: {"order[latestUploadedChapter]", "desc"},
}

func mdListManga(params url.Values) ([]SearchResult, error) {
	params.Set("limit", "24")
	params.Add("includes[]", "cover_art")
	params.Add("availableTranslatedLanguage[]", "en") // only titles with English chapters
	var res struct {
		Data []mdManga `json:"data"`
	}
	if err := httpJSON(mdURL("/manga", params), mdHeaders, &res); err != nil {
		return nil, err
	}
	out := make([]SearchResult, 0, len(res.Data))
	for i := range res.Data {
		out = append(out, mdMapManga(&res.Data[i]))
	}
	return out, nil
}

// mdRerank bubbles exact / prefix title matches up (MangaDex relevance is weak).
func mdRerank(results []SearchResult, query string) []SearchResult {
	q := strings.ToLower(strings.TrimSpace(query))
	type scored struct {
		r     SearchResult
		score int
	}
	rows := make([]scored, len(results))
	for i, r := range results {
		t := strings.ToLower(r.Title)
		score := i
		switch {
		case t == q:
			score -= 1000
		case strings.HasPrefix(t, q):
			score -= 500
		case strings.Contains(t, q):
			score -= 100
		}
		rows[i] = scored{r, score}
	}
	sort.SliceStable(rows, func(a, b int) bool { return rows[a].score < rows[b].score })
	out := make([]SearchResult, len(rows))
	for i, s := range rows {
		out[i] = s.r
	}
	return out
}

func (mangadexSource) Search(query string, opts SearchOpts) ([]SearchResult, error) {
	key := fmt.Sprintf("md:search:%s:%d:%d", query, boolInt(opts.Adult), opts.Page)
	var cached []SearchResult
	if util.ReadCache(util.SearchCache, key, 10*time.Minute, &cached) {
		return cached, nil
	}
	params := url.Values{}
	params.Set("title", query)
	params.Set("offset", strconv.Itoa(opts.Page*24))
	for _, r := range mdRatings(opts.Adult) {
		params.Add("contentRating[]", r)
	}
	params.Set("hasAvailableChapters", "true") // skip delicensed titles
	params.Set("order[relevance]", "desc")
	results, err := mdListManga(params)
	if err != nil {
		return nil, err
	}
	results = mdRerank(results, query)
	util.WriteCache(util.SearchCache, key, results)
	return results, nil
}

func (mangadexSource) Discovery(kind DiscoveryKind, page int, adult bool) ([]DiscoveryItem, error) {
	order, ok := mdOrder[kind]
	if !ok {
		order = mdOrder[Popular]
	}
	params := url.Values{}
	params.Set("offset", strconv.Itoa(page*24))
	for _, r := range mdRatings(adult) {
		params.Add("contentRating[]", r)
	}
	params.Set(order[0], order[1])
	params.Set("hasAvailableChapters", "true")
	results, err := mdListManga(params)
	if err != nil {
		return nil, err
	}
	out := make([]DiscoveryItem, len(results))
	for i, r := range results {
		out[i] = DiscoveryItem{
			ID: r.ID, Title: r.Title, Poster: r.Poster, Type: r.Type,
			IsAdult: r.IsAdult, Source: SrcMangaDex,
		}
	}
	return out, nil
}

func (mangadexSource) Filters() (Filters, error) {
	key := "md:tags"
	var cached Filters
	if util.ReadCache(util.MangaCache, key, 24*time.Hour, &cached) {
		return cached, nil
	}
	var res struct {
		Data []mdTag `json:"data"`
	}
	if err := httpJSON(mdURL("/manga/tag", nil), mdHeaders, &res); err != nil {
		return Filters{}, err
	}
	var genres []Genre
	for _, t := range res.Data {
		if t.Attributes.Group != "genre" && t.Attributes.Group != "theme" {
			continue
		}
		name := t.Attributes.Name["en"]
		if name == "" {
			for _, n := range t.Attributes.Name {
				name = n
				break
			}
		}
		genres = append(genres, Genre{ID: t.ID, Name: name})
	}
	sort.Slice(genres, func(a, b int) bool { return genres[a].Name < genres[b].Name })
	filters := Filters{
		Genres: genres,
		Tags:   []Genre{},
		Types:  []Genre{},
		Statuses: []Genre{
			{ID: "ongoing", Name: "Ongoing"},
			{ID: "completed", Name: "Completed"},
			{ID: "hiatus", Name: "Hiatus"},
			{ID: "cancelled", Name: "Cancelled"},
		},
	}
	util.WriteCache(util.MangaCache, key, filters)
	return filters, nil
}

func (mangadexSource) Browse(f BrowseFilter) ([]SearchResult, error) {
	orders := map[BrowseSort][2]string{
		SortPopular:      {"order[followedCount]", "desc"},
		SortLatest:       {"order[latestUploadedChapter]", "desc"},
		SortRating:       {"order[rating]", "desc"},
		SortAlphabetical: {"order[title]", "asc"},
	}
	order, ok := orders[f.Sort]
	if !ok {
		order = orders[SortPopular]
	}
	params := url.Values{}
	if f.GenreID != "" {
		params.Add("includedTags[]", f.GenreID)
	}
	if f.Status != "" {
		params.Add("status[]", f.Status)
	}
	for _, r := range mdRatings(f.Adult) {
		params.Add("contentRating[]", r)
	}
	params.Set("hasAvailableChapters", "true")
	params.Set(order[0], order[1])
	return mdListManga(params)
}

// mdFetchReadable paginates a manga's feed and keeps only readable
// (non-external, pages > 0) chapters.
func mdFetchReadable(mangaID string, langs []string) ([]mdChapter, error) {
	var raw []mdChapter
	offset := 0
	total := 1 << 30
	for offset < total && offset < 2000 {
		params := url.Values{}
		for _, l := range langs {
			params.Add("translatedLanguage[]", l)
		}
		params.Set("order[chapter]", "asc")
		params.Add("includes[]", "scanlation_group")
		for _, r := range []string{"safe", "suggestive", "erotica", "pornographic"} {
			params.Add("contentRating[]", r)
		}
		params.Set("limit", "100")
		params.Set("offset", strconv.Itoa(offset))
		var feed struct {
			Data  []mdChapter `json:"data"`
			Total int         `json:"total"`
		}
		if err := httpJSON(mdURL("/manga/"+mangaID+"/feed", params), mdHeaders, &feed); err != nil {
			return nil, err
		}
		raw = append(raw, feed.Data...)
		if feed.Total > 0 {
			total = feed.Total
		} else {
			total = len(raw)
		}
		offset += 100
		if len(feed.Data) == 0 {
			break
		}
	}
	out := raw[:0]
	for _, ch := range raw {
		if ch.Attributes.Pages > 0 {
			out = append(out, ch)
		}
	}
	return out, nil
}

func (mangadexSource) Info(mangaID string) (*MangaInfo, error) {
	key := "md:info:" + mangaID
	var cached MangaInfo
	if util.ReadCache(util.MangaCache, key, mdTTL, &cached) {
		return &cached, nil
	}

	params := url.Values{}
	params.Add("includes[]", "cover_art")
	var detail struct {
		Data mdManga `json:"data"`
	}
	if err := httpJSON(mdURL("/manga/"+mangaID, params), mdHeaders, &detail); err != nil {
		return nil, err
	}
	longStrip := false
	for _, t := range detail.Data.Attributes.Tags {
		if strings.EqualFold(t.Attributes.Name["en"], "long strip") {
			longStrip = true
			break
		}
	}

	// Prefer English; if it has no readable chapters, fall back to whichever
	// language has the most (so big multi-language titles still open).
	raw, err := mdFetchReadable(mangaID, []string{"en"})
	if err != nil {
		return nil, err
	}
	if len(raw) == 0 {
		var others []string
		for _, l := range detail.Data.Attributes.AvailableTranslatedLanguages {
			if l != "" && l != "en" {
				others = append(others, l)
			}
		}
		if len(others) > 0 {
			all, err := mdFetchReadable(mangaID, others)
			if err == nil {
				byLang := map[string][]mdChapter{}
				for _, ch := range all {
					l := ch.Attributes.TranslatedLanguage
					byLang[l] = append(byLang[l], ch)
				}
				for _, arr := range byLang {
					if len(arr) > len(raw) {
						raw = arr
					}
				}
			}
		}
	}

	// One release per chapter number (first scanlation wins).
	seen := map[string]bool{}
	var picked []mdChapter
	for _, ch := range raw {
		k := "oneshot:" + ch.ID
		if ch.Attributes.Chapter != nil {
			k = *ch.Attributes.Chapter
		}
		if seen[k] {
			continue
		}
		seen[k] = true
		picked = append(picked, ch)
	}
	chNum := func(ch mdChapter) float64 {
		if ch.Attributes.Chapter == nil {
			return 1e9
		}
		n, err := strconv.ParseFloat(*ch.Attributes.Chapter, 64)
		if err != nil {
			return 1e9
		}
		return n
	}
	sort.SliceStable(picked, func(a, b int) bool { return chNum(picked[a]) < chNum(picked[b]) })

	info := MangaInfo{
		ID:         mangaID,
		Title:      mdPickTitle(&detail.Data),
		Type:       "Manga",
		ForceStrip: longStrip,
	}
	for i, ch := range picked {
		title := ""
		if ch.Attributes.Title != nil {
			title = *ch.Attributes.Title
		}
		if title == "" {
			if ch.Attributes.Chapter != nil {
				title = "Chapter " + *ch.Attributes.Chapter
			} else {
				title = "Oneshot"
			}
		}
		num := chNum(ch)
		if num >= 1e9 {
			num = float64(i + 1)
		}
		scan := ""
		for _, r := range ch.Relationships {
			if r.Type == "scanlation_group" {
				scan = r.ID
				break
			}
		}
		created := int64(0)
		for _, ts := range []string{ch.Attributes.PublishAt, ch.Attributes.CreatedAt} {
			if ts == "" {
				continue
			}
			if t, err := time.Parse(time.RFC3339, ts); err == nil {
				created = t.UnixMilli()
				break
			}
		}
		info.Chapters = append(info.Chapters, Chapter{
			ID: ch.ID, Title: title, Number: num, Index: i,
			PageCount: ch.Attributes.Pages, ScanID: scan, CreatedAt: created,
		})
	}
	util.WriteCache(util.MangaCache, key, info)
	return &info, nil
}

func (mangadexSource) Pages(_ string, chapterID string) (*ReadChapter, error) {
	key := "md:pages:" + chapterID
	var cached ReadChapter
	if util.ReadCache(util.ChaptersCache, key, mdTTL, &cached) {
		return &cached, nil
	}
	var res struct {
		BaseURL string `json:"baseUrl"`
		Chapter struct {
			Hash string   `json:"hash"`
			Data []string `json:"data"`
		} `json:"chapter"`
	}
	if err := httpJSON(mdURL("/at-home/server/"+chapterID, nil), mdHeaders, &res); err != nil {
		return nil, err
	}
	out := ReadChapter{ID: chapterID}
	for i, file := range res.Chapter.Data {
		out.Pages = append(out.Pages, Page{
			ID:          fmt.Sprintf("%s-%d", chapterID, i),
			URL:         fmt.Sprintf("%s/data/%s/%s", res.BaseURL, res.Chapter.Hash, file),
			Number:      i,
			AspectRatio: 1,
		})
	}
	util.WriteCache(util.ChaptersCache, key, out)
	return &out, nil
}

// MangaDex has no public "similar" endpoint; recommendations fall back elsewhere.
func (mangadexSource) Related(string, int, bool) ([]DiscoveryItem, error) {
	return []DiscoveryItem{}, nil
}

func boolInt(b bool) int {
	if b {
		return 1
	}
	return 0
}
